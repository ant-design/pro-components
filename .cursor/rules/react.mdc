---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
