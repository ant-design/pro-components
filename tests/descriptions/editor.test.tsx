import type {
  ProDescriptionsActionType,
  ProDescriptionsItemProps,
} from '@ant-design/pro-descriptions';
import Descriptions from '@ant-design/pro-descriptions';
import type { RowEditableConfig } from '@ant-design/pro-utils';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import { Form, InputNumber } from 'antd';
import useMergedState from 'rc-util/es/hooks/useMergedState';
import React, { act, useRef } from 'react';
type DataSourceType = {
  id: number;
  title?: string;
  labels?: {
    name: string;
    color: string;
  }[];
  state?: string;
  time?: {
    created_at?: number;
  };
  children?: DataSourceType;
};

const defaultData: DataSourceType = {
  id: 624748504,
  title: 'üêõ [BUG]yarn installÂëΩ‰ª§ antd2.4.5‰ºöÊä•Èîô',
  labels: [{ name: 'bug', color: 'error' }],
  time: {
    created_at: 1590486176000,
  },
  state: 'processing',
};

const columns: ProDescriptionsItemProps<DataSourceType>[] = [
  {
    dataIndex: 'index',
    valueType: 'indexBorder',
    renderFormItem: () => <InputNumber />,
  },
  {
    title: 'Ê†áÈ¢ò',
    dataIndex: 'title',
    copyable: true,
    ellipsis: true,
    tooltip: 'Ê†áÈ¢òËøáÈïø‰ºöËá™Âä®Êî∂Áº©',
    formItemProps: {
      rules: [
        {
          required: true,
          message: 'Ê≠§È°π‰∏∫ÂøÖÂ°´È°π',
        },
      ],
    },
  },
  {
    title: 'Áä∂ÊÄÅ',
    dataIndex: 'state',
    valueType: 'select',
    valueEnum: {
      all: { text: 'ÂÖ®ÈÉ®', status: 'Default' },
      open: {
        text: 'Êú™Ëß£ÂÜ≥',
        status: 'Error',
      },
      closed: {
        text: 'Â∑≤Ëß£ÂÜ≥',
        status: 'Success',
      },
      processing: {
        text: 'Ëß£ÂÜ≥‰∏≠',
        status: 'Processing',
      },
    },
  },
  {
    title: 'ÂàõÂª∫Êó∂Èó¥',
    dataIndex: ['time', 'created_at'],
  },
];

const DescriptionsDemo = (
  props: {
    type?: 'multiple';
    defaultKeys?: React.Key[];
    editorRowKeys?: React.Key[];
    onEditorChange?: (editorRowKeys: React.Key[]) => void;
    dataSource?: DataSourceType;
    onDataSourceChange?: (dataSource: DataSourceType) => void;
  } & RowEditableConfig<DataSourceType>,
) => {
  const [form] = Form.useForm();
  const actionRef = useRef<ProDescriptionsActionType>();
  const [editableKeys, setEditorRowKeys] = useMergedState<React.Key[]>(
    () => props.defaultKeys || [],
    {
      value: props.editorRowKeys,
      onChange: props.onEditorChange,
    },
  );
  const [dataSource, setDataSource] = useMergedState<
    DataSourceType,
    DataSourceType
  >(props.dataSource as any, {
    value: props.dataSource,
    onChange: props.onDataSourceChange,
  });
  return (
    <Descriptions<DataSourceType>
      columns={columns}
      actionRef={actionRef}
      request={async () => ({
        data: defaultData,
        total: 3,
        success: true,
      })}
      title={
        <a
          id="reset_test"
          onClick={() => {
            form.resetFields();
          }}
        >
          ÈáçÁΩÆ
        </a>
      }
      dataSource={dataSource}
      onDataSourceChange={setDataSource}
      editable={{
        ...props,
        form,
        type: props.type,
        editableKeys,
        onSave: props.onSave,
        onChange: (keys) => setEditorRowKeys(keys),
      }}
    />
  );
};

afterEach(() => {
  cleanup();
});

describe('Descriptions', () => {
  afterEach(() => {
    cleanup();
  });

  it('üìù Descriptions close editable', async () => {
    const wrapper = render(
      <Descriptions<DataSourceType>
        title="Âü∫Êú¨‰ΩøÁî®"
        columns={columns}
        dataSource={defaultData}
      />,
    );

    await wrapper.findAllByText('Âü∫Êú¨‰ΩøÁî®');
    expect(!!wrapper.baseElement.querySelector('.anticon-edit')).toBeFalsy();
  });

  it('üìù Descriptions support editable', async () => {
    const wrapper = render(
      <Descriptions<DataSourceType>
        title="Âü∫Êú¨‰ΩøÁî®"
        columns={columns}
        dataSource={defaultData}
        editable={{}}
      />,
    );
    await wrapper.findAllByText('Âü∫Êú¨‰ΩøÁî®');
    expect(!!wrapper.baseElement.querySelector('.anticon-edit')).toBeTruthy();
  });

  it('üìù support onEditorChange', async () => {
    const fn = vi.fn();
    const wrapper = render(
      <DescriptionsDemo
        onEditorChange={(keys) => {
          fn(keys);
        }}
      />,
    );
    await wrapper.findAllByText('ÈáçÁΩÆ');

    act(() => {
      wrapper.baseElement
        .querySelectorAll<HTMLSpanElement>('span.anticon-edit')[0]
        ?.click();
    });

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith(['title']);
    });
  });

  it('üìù support set Form', async () => {
    const wrapper = render(<DescriptionsDemo editorRowKeys={['title']} />);

    await wrapper.findAllByText('ÈáçÁΩÆ');

    act(() => {
      fireEvent.change(
        wrapper.baseElement
          .querySelectorAll<HTMLSpanElement>(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[0]
          .querySelectorAll('.ant-input')[0],
        { target: { value: 'test' } },
      );
    });
    await waitFor(() => {
      expect(wrapper.queryByDisplayValue('test')).toBeTruthy();
    });

    act(() => {
      wrapper.queryByText('ÈáçÁΩÆ')?.click();
    });
    await waitFor(() => {
      expect(
        wrapper.queryByDisplayValue('üêõ [BUG]yarn installÂëΩ‰ª§ antd2.4.5‰ºöÊä•Èîô'),
      ).toBeTruthy();
    });
  });

  it('üìù renderFormItem run defaultRender', async () => {
    const wrapper = render(
      <Descriptions<DataSourceType>
        editable={{
          editableKeys: ['title'],
        }}
        columns={[
          {
            dataIndex: 'title',
            renderFormItem: (item, config) => {
              return config.defaultRender(item);
            },
          },
        ]}
        dataSource={defaultData}
      />,
    );
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('üìù columns support editable test', async () => {
    const wrapper = render(
      <Descriptions
        editable={{
          editableKeys: ['title'],
        }}
        columns={[
          {
            dataIndex: 'title',
            editable: (text, record, index) => {
              return index === 1;
            },
          },
          {
            dataIndex: 'title2',
            editable: false,
          },
        ]}
        dataSource={defaultData}
      />,
    );
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('üìù support actionRender', async () => {
    const wrapper = render(
      <Descriptions
        editable={{
          editableKeys: ['title'],
          actionRender: () => [
            <div key="test" id="test">
              xx
            </div>,
          ],
        }}
        columns={[
          {
            dataIndex: 'title',
            editable: (text, record, index) => {
              return index === 1;
            },
          },
          {
            dataIndex: 'title2',
            editable: false,
          },
        ]}
        dataSource={defaultData}
      />,
    );
    expect(!!wrapper.queryByText('xx')).toBe(true);
  });

  it('üìù support editorRowKeys', async () => {
    const wrapper = render(<DescriptionsDemo editorRowKeys={['title']} />);

    await wrapper.findAllByDisplayValue(
      'üêõ [BUG]yarn installÂëΩ‰ª§ antd2.4.5‰ºöÊä•Èîô',
    );
    // Á¨¨‰∏ÄË°åÂ∫îËØ•ÁºñËæëÊÄÅ
    expect(
      wrapper.baseElement
        .querySelectorAll(
          'td.ant-descriptions-item .ant-descriptions-item-content',
        )[0]
        .querySelectorAll('input').length > 0,
    ).toBeTruthy();

    // Á¨¨‰∫åË°å‰∏çÂ∫îËØ•ÊòØÁºñËæëÊÄÅ
    expect(
      wrapper.baseElement
        .querySelectorAll(
          'td.ant-descriptions-item .ant-descriptions-item-content',
        )[1]
        .querySelectorAll('input').length > 0,
    ).toBeFalsy();
  });

  it('üìù support cancel click', async () => {
    const fn = vi.fn();
    const wrapper = render(
      <DescriptionsDemo
        onEditorChange={(keys) => {
          fn(keys);
        }}
      />,
    );
    await wrapper.findAllByText('ÈáçÁΩÆ');
    act(() => {
      wrapper.baseElement
        .querySelector<HTMLDivElement>('span.anticon-edit')
        ?.click();
    });
    await waitFor(() => {
      expect(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[0]
          .querySelectorAll('input').length > 0,
      ).toBeTruthy();
    });

    act(() => {
      wrapper.baseElement
        .querySelectorAll(
          'td.ant-descriptions-item .ant-descriptions-item-content',
        )[0]
        .querySelector<HTMLSpanElement>(`span.anticon-close`)
        ?.click();
    });
    await waitFor(() => {
      expect(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[0]
          .querySelectorAll('input').length > 0,
      ).toBeFalsy();
    });
  });

  it('üìù support cancel click render false', async () => {
    const fn = vi.fn();
    const wrapper = render(
      <DescriptionsDemo
        onEditorChange={(keys) => {
          fn(keys);
        }}
        onCancel={async () => false}
      />,
    );
    await wrapper.findAllByText('ÈáçÁΩÆ');
    act(() => {
      wrapper.baseElement
        .querySelector<HTMLSpanElement>('span.anticon-edit')
        ?.click();
    });
    await waitFor(() => {
      expect(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[0]
          .querySelectorAll('input').length > 0,
      ).toBeTruthy();
    });

    act(() => {
      wrapper.baseElement
        .querySelector<HTMLSpanElement>(
          'td.ant-descriptions-item .ant-descriptions-item-content',
        )
        ?.querySelector<HTMLSpanElement>(`span.anticon-close`)
        ?.click();
    });

    await waitFor(() => {
      expect(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[0]
          .querySelectorAll('input').length > 0,
      ).toBeFalsy();
    });
  });

  it('üìù type=single, only edit one rows', async () => {
    const fn = vi.fn();
    const wrapper = render(
      <DescriptionsDemo
        defaultKeys={['state']}
        onEditorChange={(keys) => {
          fn(keys);
        }}
      />,
    );
    await wrapper.findAllByText('ÈáçÁΩÆ');
    act(() => {
      wrapper.baseElement
        .querySelector<HTMLSpanElement>('span.anticon-edit')
        ?.click();
    });

    await waitFor(() => {
      expect(fn).not.toBeCalled();
    });
  });

  it('üìù type=multiple, edit multiple rows', async () => {
    const fn = vi.fn();
    const wrapper = render(
      <DescriptionsDemo
        type="multiple"
        defaultKeys={['state']}
        onEditorChange={(keys) => {
          fn(keys);
        }}
      />,
    );
    await wrapper.findAllByText('ÈáçÁΩÆ');
    act(() => {
      wrapper.baseElement
        .querySelector<HTMLSpanElement>('span.anticon-edit')
        ?.click();
    });
    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith(['state', 'title']);
    });
  });

  it('üìù support onSave', async () => {
    const fn = vi.fn();
    const wrapper = render(<DescriptionsDemo onSave={(key) => fn(key)} />);
    await wrapper.findAllByText('ÈáçÁΩÆ');
    act(() => {
      wrapper.baseElement
        .querySelectorAll<HTMLSpanElement>('span.anticon-edit')[1]
        ?.click();
    });

    await waitFor(() => {
      expect(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[1]
          .querySelectorAll('input').length > 0,
      ).toBeTruthy();
    });

    act(() => {
      wrapper.baseElement
        .querySelectorAll(
          'td.ant-descriptions-item .ant-descriptions-item-content',
        )[1]
        .querySelector<HTMLSpanElement>('span.anticon-check')
        ?.click();
    });
    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith('state');
    });
  });

  it('üìù support onSave support false', async () => {
    const fn = vi.fn();
    const wrapper = render(
      <DescriptionsDemo
        onSave={async (key) => {
          fn(key);
          return false;
        }}
      />,
    );

    await wrapper.findAllByText('ÈáçÁΩÆ');

    act(() => {
      wrapper.baseElement
        .querySelectorAll<HTMLSpanElement>('span.anticon-edit')[1]
        ?.click();
    });

    await waitFor(() => {
      expect(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[1]
          .querySelectorAll('input').length > 0,
      ).toBeTruthy();
    });

    act(() => {
      wrapper.baseElement
        .querySelectorAll(
          'td.ant-descriptions-item .ant-descriptions-item-content',
        )[1]
        .querySelector<HTMLSpanElement>(`span.anticon-check`)
        ?.click();
    });

    await waitFor(() => {
      expect(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[1]
          .querySelectorAll('input').length > 0,
      ).toBeTruthy();
    });

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith('state');
    });
  });

  it('üìù support onCancel', async () => {
    const fn = vi.fn();
    const wrapper = render(<DescriptionsDemo onCancel={(key) => fn(key)} />);

    await wrapper.findAllByText('ÈáçÁΩÆ');

    act(() => {
      wrapper.baseElement
        .querySelectorAll<HTMLSpanElement>('span.anticon-edit')[1]
        ?.click();
    });

    await waitFor(() => {
      expect(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[1]
          .querySelectorAll('input').length > 0,
      ).toBeTruthy();
    });
    act(() => {
      wrapper.baseElement
        .querySelectorAll(
          'td.ant-descriptions-item .ant-descriptions-item-content',
        )[1]
        .querySelector<HTMLSpanElement>(`span.anticon-close`)
        ?.click();
    });

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith('state');
    });
  });

  it('üìù support form rules', async () => {
    const fn = vi.fn();
    const wrapper = render(
      <DescriptionsDemo onSave={(key, row) => fn(row.title)} />,
    );

    await wrapper.findAllByText('ÈáçÁΩÆ');

    act(() => {
      wrapper.baseElement
        .querySelectorAll<HTMLSpanElement>('span.anticon-edit')[0]
        ?.click();
    });

    await waitFor(() => {
      expect(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[0]
          .querySelectorAll('input').length > 0,
      ).toBeTruthy();
    });

    act(() => {
      fireEvent.change(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[0]
          .querySelectorAll('input')![0],
        {
          target: {
            value: '',
          },
        },
      );
    });
    act(() => {
      wrapper.baseElement
        .querySelectorAll(
          'td.ant-descriptions-item .ant-descriptions-item-content',
        )[0]
        .querySelectorAll<HTMLSpanElement>(`span.anticon-check`)[0]
        .click();
    });

    await waitFor(() => {
      // Ê≤°ÊúâÈÄöËøáÈ™åËØÅÔºå‰∏çËß¶Âèë onSave
      expect(fn).not.toBeCalled();
    });

    act(() => {
      fireEvent.change(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[0]
          .querySelectorAll('input')![0],
        {
          target: {
            value: 'qixian',
          },
        },
      );
    });

    act(() => {
      fireEvent.click(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[0]
          .querySelector('span.anticon-check')!,
        {},
      );
    });

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith('qixian');
    });
  });

  it('üìù when dataIndex is array', async () => {
    const fn = vi.fn();
    const wrapper = render(
      <DescriptionsDemo onSave={(key, row) => fn(row?.time?.created_at)} />,
    );
    await wrapper.findAllByText('ÈáçÁΩÆ');

    act(() => {
      wrapper.baseElement
        .querySelectorAll<HTMLSpanElement>('span.anticon-edit')[2]
        ?.click();
    });

    act(() => {
      fireEvent.change(
        wrapper.baseElement
          .querySelectorAll(
            'td.ant-descriptions-item .ant-descriptions-item-content',
          )[2]
          .querySelector(`input.ant-input`)!,
        {
          target: {
            value: '2021-05-26 09:42:56',
          },
        },
      );
    });

    act(() => {
      wrapper.baseElement
        .querySelectorAll(
          'td.ant-descriptions-item .ant-descriptions-item-content',
        )[2]
        .querySelectorAll<HTMLDivElement>(`span.anticon-check`)[0]
        ?.click();
    });

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith('2021-05-26 09:42:56');
    });
  });
});
