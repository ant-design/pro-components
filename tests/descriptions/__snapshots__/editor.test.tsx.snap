// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Descriptions > üìù columns support editable test 1`] = `
<DocumentFragment>
  <div
    class="ant-descriptions ant-pro-descriptions"
  >
    <div
      class="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            class="ant-descriptions-row"
          >
            <td
              class="ant-descriptions-item"
              colspan="1"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-content"
                  style="min-width: 0;"
                >
                  <div
                    style="margin: -5px 0px;"
                  >
                    <div
                      style="display: flex; gap: 8px; align-items: baseline;"
                    >
                      <div
                        class="ant-form-item"
                        style="margin-block-start: -5px; margin-block-end: -5px; margin-inline-start: 0; margin-inline-end: 0; margin: 0px;"
                      >
                        <div
                          class="ant-row ant-form-item-row"
                        >
                          <div
                            class="ant-col ant-form-item-control"
                          >
                            <div
                              class="ant-form-item-control-input"
                            >
                              <div
                                class="ant-form-item-control-input-content"
                              >
                                <span
                                  class="ant-input-affix-wrapper ant-input-outlined"
                                >
                                  <input
                                    class="ant-input"
                                    placeholder="ËØ∑ËæìÂÖ•"
                                    type="text"
                                    value="üêõ [BUG]yarn installÂëΩ‰ª§ antd2.4.5‰ºöÊä•Èîô"
                                  />
                                  <span
                                    class="ant-input-suffix"
                                  >
                                    <span
                                      class="ant-input-clear-icon"
                                      role="button"
                                      tabindex="-1"
                                    >
                                      <span
                                        aria-label="close-circle"
                                        class="anticon anticon-close-circle"
                                        role="img"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          data-icon="close-circle"
                                          fill="currentColor"
                                          fill-rule="evenodd"
                                          focusable="false"
                                          height="1em"
                                          viewBox="64 64 896 896"
                                          width="1em"
                                        >
                                          <path
                                            d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </span>
                                </span>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        style="display: flex; max-height: 32px; align-items: center; gap: 8px;"
                      >
                        <a>
                          <span
                            aria-label="check"
                            class="anticon anticon-check"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="check"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 00-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
                              />
                            </svg>
                          </span>
                        </a>
                        <a>
                          <span
                            aria-label="close"
                            class="anticon anticon-close"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="close"
                              fill="currentColor"
                              fill-rule="evenodd"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                              />
                            </svg>
                          </span>
                        </a>
                      </div>
                    </div>
                  </div>
                </span>
              </div>
            </td>
            <td
              class="ant-descriptions-item"
              colspan="2"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-content"
                  style="min-width: 0;"
                >
                  -
                </span>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Descriptions > üìù formItemRender run defaultRender 1`] = `
<DocumentFragment>
  <div
    class="ant-descriptions ant-pro-descriptions"
  >
    <div
      class="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            class="ant-descriptions-row"
          >
            <td
              class="ant-descriptions-item"
              colspan="3"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-content"
                  style="min-width: 0;"
                >
                  <div
                    style="margin: -5px 0px;"
                  >
                    <div
                      style="display: flex; gap: 8px; align-items: baseline;"
                    >
                      <div
                        class="ant-form-item"
                        style="margin-block-start: -5px; margin-block-end: -5px; margin-inline-start: 0; margin-inline-end: 0; margin: 0px;"
                      >
                        <div
                          class="ant-row ant-form-item-row"
                        >
                          <div
                            class="ant-col ant-form-item-control"
                          >
                            <div
                              class="ant-form-item-control-input"
                            >
                              <div
                                class="ant-form-item-control-input-content"
                              >
                                <span
                                  class="ant-input-affix-wrapper ant-input-outlined"
                                >
                                  <input
                                    class="ant-input"
                                    placeholder="ËØ∑ËæìÂÖ•"
                                    type="text"
                                    value="üêõ [BUG]yarn installÂëΩ‰ª§ antd2.4.5‰ºöÊä•Èîô"
                                  />
                                  <span
                                    class="ant-input-suffix"
                                  >
                                    <span
                                      class="ant-input-clear-icon"
                                      role="button"
                                      tabindex="-1"
                                    >
                                      <span
                                        aria-label="close-circle"
                                        class="anticon anticon-close-circle"
                                        role="img"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          data-icon="close-circle"
                                          fill="currentColor"
                                          fill-rule="evenodd"
                                          focusable="false"
                                          height="1em"
                                          viewBox="64 64 896 896"
                                          width="1em"
                                        >
                                          <path
                                            d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </span>
                                </span>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                      <div
                        style="display: flex; max-height: 32px; align-items: center; gap: 8px;"
                      >
                        <a>
                          <span
                            aria-label="check"
                            class="anticon anticon-check"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="check"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 00-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
                              />
                            </svg>
                          </span>
                        </a>
                        <a>
                          <span
                            aria-label="close"
                            class="anticon anticon-close"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="close"
                              fill="currentColor"
                              fill-rule="evenodd"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M799.86 166.31c.02 0 .04.02.08.06l57.69 57.7c.04.03.05.05.06.08a.12.12 0 010 .06c0 .03-.02.05-.06.09L569.93 512l287.7 287.7c.04.04.05.06.06.09a.12.12 0 010 .07c0 .02-.02.04-.06.08l-57.7 57.69c-.03.04-.05.05-.07.06a.12.12 0 01-.07 0c-.03 0-.05-.02-.09-.06L512 569.93l-287.7 287.7c-.04.04-.06.05-.09.06a.12.12 0 01-.07 0c-.02 0-.04-.02-.08-.06l-57.69-57.7c-.04-.03-.05-.05-.06-.07a.12.12 0 010-.07c0-.03.02-.05.06-.09L454.07 512l-287.7-287.7c-.04-.04-.05-.06-.06-.09a.12.12 0 010-.07c0-.02.02-.04.06-.08l57.7-57.69c.03-.04.05-.05.07-.06a.12.12 0 01.07 0c.03 0 .05.02.09.06L512 454.07l287.7-287.7c.04-.04.06-.05.09-.06a.12.12 0 01.07 0z"
                              />
                            </svg>
                          </span>
                        </a>
                      </div>
                    </div>
                  </div>
                </span>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
